src/database/test/DatabaseTests.cpp

```diff
@@ -349,3 +355,430 @@ TEST_CASE("schema test", "[db]")
+                    << __func__ << ": exception " << e.what()
+                    << " while calling pre-upgrade function";
+                throw;
+            }
+        }
+    }
+};
+
+// This tests upgrading from MIN_SCHEMA_VERSION to SCHEMA_VERSION by creating
+// ledger entries with the old MIN_SCHEMA_VERSION after database creation, then
+// validating their contents after the upgrade to SCHEMA_VERSION.
+TEST_CASE("schema upgrade test", "[db]")
+{
+    using OptLiabilities = stellar::optional<Liabilities>;
+    using AccOptLiabilitiesPair = std::pair<AccountEntry, OptLiabilities>;
+    using AccOptLiabilitiesVec = std::vector<AccOptLiabilitiesPair>;
+    using TLOptLiabilitiesPair = std::pair<TrustLineEntry, OptLiabilities>;
+    using TLOptLiabilitiesVec = std::vector<TLOptLiabilitiesPair>;
+
+    auto addOneOldSchemaAccount = [](SchemaUpgradeTestApplication& app,
+                                     AccOptLiabilitiesPair const& aol) {
+        auto const& ae = aol.first;
+        auto const& liabilities = aol.second;
+        auto& session = app.getDatabase().getSession();
+        auto accountIDStr = KeyUtils::toStrKey<PublicKey>(ae.accountID);
+        auto inflationDestStr =
+            ae.inflationDest ? KeyUtils::toStrKey<PublicKey>(*ae.inflationDest)
+                             : "";
+        auto inflationDestInd = ae.inflationDest ? soci::i_ok : soci::i_null;
+        auto homeDomainStr = decoder::encode_b64(ae.homeDomain);
+        auto signersStr = decoder::encode_b64(xdr::xdr_to_opaque(ae.signers));
+        auto thresholdsStr = decoder::encode_b64(ae.thresholds);
+
+        soci::transaction tx(session);
+
+        // Use raw SQL to perform database operations, since we're writing in an
+        // old database format, and calling standard interfaces to create
+        // accounts or trustlines would use SQL corresponding to the new format
+        // to which we'll soon upgrade.
+        session << "INSERT INTO accounts ( "
+                   "accountid, balance, seqnum, numsubentries, inflationdest,"
+                   "homedomain, thresholds, signers, flags, lastmodified, "
+                   "buyingliabilities, sellingliabilities "
+                   ") VALUES ( "
+                   ":id, :v1, :v2, :v3, :v4, :v5, :v6, :v7, :v8, :v9, :v10, "
+                   ":v11 "
+                   ")",
+            soci::use(accountIDStr), soci::use(ae.balance),
+            soci::use(ae.seqNum), soci::use(ae.numSubEntries),
+            soci::use(inflationDestStr, inflationDestInd),
+            soci::use(homeDomainStr), soci::use(thresholdsStr),
+            soci::use(signersStr), soci::use(ae.flags),
+            soci::use(app.getLedgerManager().getLastClosedLedgerNum()),
+            soci::use(liabilities ? liabilities->buying : 0),
+            soci::use(liabilities ? liabilities->selling : 0);
+
+        if (!liabilities)
+        {
+            session << "UPDATE accounts SET buyingliabilities = NULL, "
```

Comment:
Why not just use indicators on line 439, 440 instead of this?

Reply:
I couldn't figure out how to get it to typecheck!  The compiler was telling me that I was trying to convert a ```size_t``` into an ```indicator``` for the second argument, and I was baffled as to why (```soci::i_ok``` sure looks like an ```indicator``` to me).

Since you asked, I went back and fought with the compiler for a while longer.  Eventually I figured out that the problem was that the overload that I was _actually_ trying to use declares the input ```indicator``` as a _non-_```const``` reference.  It did not occur to me that it might do that, since on an update, I figured, we were only reading from the indicator parameter, not modifying it.

So, I made a non-```const``` local to hold ```soci::i_ok``` (or ```soci::i_null```, as appropriate), and with that I've been able to implement your suggestion.